src/request/swap-iframe/SwapIFrameApi.js

```diff
@@ -0,0 +1,385 @@
+            inputs.sort((a, b) => {
+                if (a.hash !== b.hash) return a.hash < b.hash ? -1 : 1;
+                return a.index - b.index;
+            });
+
+            // Construct outputs
+            const outputs = [{
+                address: parsedRequest.fund.htlcAddress,
+                value: storedRequest.fund.recipientOutput.value,
+            }];
+
+            // Add change output
+            if (storedRequest.fund.changeOutput) {
+                // The address is set in SignSwap after password entry.
+                outputs.push(/** @type {{address: string, value: number}} */ (storedRequest.fund.changeOutput));
+            }
+
+            // Sort outputs by value ASC, then address ASC
+            outputs.sort((a, b) => (a.value - b.value) || (a.address < b.address ? -1 : 1));
+
+            // Construct transaction
+            const psbt = new BitcoinJS.Psbt({ network: BitcoinUtils.Network });
+
+            // Add inputs
+            // @ts-ignore Argument of type 'Uint8Array' is not assignable to parameter of type 'Buffer'.
+            psbt.addInputs(inputs);
+            // Add outputs
+            psbt.addOutputs(outputs);
+
+            // Sign
+            const keyPairs = privateKeys.btc.map(privateKey => BitcoinJS.ECPair.fromPrivateKey(
+                // @ts-ignore Argument of type 'import("...").Buffer' is not assignable to parameter of type 'Buffer'.
+                BitcoinJS.Buffer.from(privateKey, 'hex'),
+            ));
+            for (const keyPair of keyPairs) {
+                psbt.signAllInputs(keyPair);
+            }
+
+            // Verify that all inputs are signed
+            if (!psbt.validateSignaturesOfAllInputs()) {
+                throw new Error('Invalid or missing signature(s) for BTC transaction.');
+            }
+
+            // Finalize
+            psbt.finalizeAllInputs();
+
+            // Extract tx
+            const tx = psbt.extractTransaction();
+
+            /** @type {KeyguardRequest.SignedBitcoinTransaction} */
+            result.btc = {
+                transactionHash: tx.getId(),
+                raw: tx.toHex(),
+            };
+        }
+
+        if (parsedRequest.redeem.type === 'NIM' && storedRequest.redeem.type === 'NIM') {
+            await loadNimiq();
+
+            const privateKey = new Nimiq.PrivateKey(Nimiq.BufferUtils.fromHex(privateKeys.nim));
+            const publicKey = Nimiq.PublicKey.derive(privateKey);
+
+            const transaction = Nimiq.Transaction.fromPlain({
+                ...storedRequest.redeem.transaction.toPlain(),
+                sender: parsedRequest.redeem.htlcAddress,
+            });
+
+            const signature = Nimiq.Signature.create(privateKey, publicKey, transaction.serializeContent());
+
+            /** @type {KeyguardRequest.SignatureResult} */
+            result.nim = {
+                publicKey: publicKey.serialize(),
+                signature: signature.serialize(),
+            };
```

Comment:
I was wondering why the transaction proof including the pre-image is not constructed in the KG already?

And in the hub, currently a constant pre-image seems to be used.
How does that work?